[C#] New Concept Exercise: arrays (#1257)
exercism/v3: languages/csharp/exercises/concept/arrays/.docs/instructions.md

> In this exercise, you're an avid bird watcher that keeps track of how many birds have visited your garden in the last seven days.

Reviewer: ```suggestion
You're an avid bird watcher that keeps track of how many birds have visited your garden in the last seven days.
```

You have five tasks, all dealing with the numbers of birds that visited your garden.

### 1. Calculate the total number of visiting birds

Implement the `BirdCount.Total()` method to return the total number of birds that have visited your garden:

```csharp

> int[] dayCounts = { 2, 5, 1, 4, 7, 0 };

Reviewer 2: The `dayCounts` array name feels... clunky 🤷‍♂️. Would `dailyCount` or `birdsPerDay` indicate the purpose better?

Reviewer: I like `birdsPerDay`!

Author: I like it too!

var birdCount = new BirdCount(dayCounts);
birdCount.Total();
// => 19
```

### 2. Calculate the number of busy days

Some days are busier that others. A busy day is one where five or more birds have visited your garden.
Implement the `BirdCount.BusyDays()` method to return the number of busy days:

```csharp
int[] dayCounts = { 2, 5, 1, 4, 7, 0 };
var birdCount = new BirdCount(dayCounts);
birdCount.BusyDays();
// => 2
```

### 3. Check how many birds visited yesterday

> Implement the `BirdCount.Yesterday()` method to return how many birds visited your garden yesterday, which is the second last element:

Reviewer: ```suggestion
Implement the `BirdCount.Yesterday()` method to return how many birds visited your garden yesterday, which is the penultimate element:
```

Author: Ah. I did some googling and found some things on "second last". I do think "penultimate" will be unfamiliar to many non-native speakers. What about:

```suggestion
Implement the `BirdCount.Yesterday()` method to return how many birds visited your garden yesterday, which is the sixth element:
```

Reviewer: I think indexing from the right like you had is better. How about "last but one"?

Reviewer 2: This is a tricky one to describe in this manner to non-native speakers (or even native speakers from different regions). Also, reading until this point I hadn't considered how the ordering would related to the current day. Perhaps it could be explained earlier in the text that the array is ordered so that today's count is the last, or right-most, count in the array?

Author: Yes, let me explain the ordering a bit. I'll rephrase.

> Implement the `BirdCount.Yesterday()` method to return how many birds visited your garden yesterday. The bird counts are ordered by day, with the first element being the count of six days ago, and the last element being today's count.

Reviewer: I think the "six days ago" adds noise, which makes it harder to comprehend. I think it's a good solution other than that though :)

```suggestion
Implement the `BirdCount.Yesterday()` method to return how many birds visited your garden yesterday. The bird counts are ordered by day, with the first element being the count of the oldest day, and the last element being today's count.
```

Reviewer: I also like that this doesn't give the answer away as much

Author: Yes, good idea!

```csharp
int[] dayCounts = { 2, 4, 1, 6, 7, 0 };
var birdCount = new BirdCount(dayCounts);
birdCount.Yesterday();
// => 7
```

> ### 4. Check for a specific count

Reviewer: ```suggestion
### 4. Check the count for a specific day
```

> Implement the `BirdCount.HasDayWithoutBirds()` method that returns `true` if there was a day at which no birds (count is zero) visited the garden; otherwise, return `false`:

Reviewer: ```suggestion
Implement the `BirdCount.HasDayWithoutBirds()` method that returns `true` if there was a day at which zero birds visited the garden; otherwise, return `false`:
```

> Your neighbor is also a bird enthusiast. Occasionally, your neighbor wants to know if your garden was also visited by a specific number of birds. Implement the `BirdCount.Exact()` method that the number of birds to check as a parameter and returns a boolean that indicates if that exact number of birds has visited your garden:

Reviewer: ```suggestion
Your neighbor is also a bird enthusiast. Occasionally, she wants to know if your garden was also visited by a specific number of birds. Implement the `BirdCount.Exact()` method that the number of birds to check as a parameter and returns a boolean that indicates if that exact number of birds has visited your garden:
```

Reviewer: I don't understand this one :) It feels like it should take two params - one for the day and one for the count on that day.

Author: Ah yes, you are right. That does make more sense. I'll add it.

Author: Well that would not work for what I'm trying to use the method for: getting students to use `Array.IndexOf`. Let me see if I can rephrase.

Author: So my goal is to have the student know about the `Array` class and its static methods: https://docs.microsoft.com/en-us/dotnet/api/system.array?view=netframework-4.8#methods Unfortunately, most of those methods are either complex or require complex parameters (predicates), which I don't want to use.
